File: src/omnia/data/structure/tuple/Nonuplet.java
package omnia.data.structure.tuple;

import java.util.function.Function;

public interface Nonuplet<T> extends Nonuple<T, T, T, T, T, T, T, T, T>, Tuples.AtMostNonuplet<T> {

  static <T> Nonuplet<T> of(T first, T second, T third, T fourth, T fifth, T sixth, T seventh, T eighth, T ninth) {
    return new ImmutableNonuplet<>(first, second, third, fourth, fifth, sixth, seventh, eighth, ninth);
  }

  @Override
  <R> Nonuplet<R> map(Function<? super T, ? extends R> mapper);

  @Override
  Octuplet<T> dropFirst();

  @Override
  Octuplet<T> dropSecond();

  @Override
  Octuplet<T> dropThird();

  @Override
  Octuplet<T> dropFourth();

  @Override
  Octuplet<T> dropFifth();

  @Override
  Octuplet<T> dropSixth();

  @Override
  Octuplet<T> dropSeventh();

  @Override
  Octuplet<T> dropEighth();

  @Override
  Octuplet<T> dropNinth();

  @Override
  Decuplet<T> concat(T object);
}

File: src/omnia/data/structure/tuple/Nonuplet.kt
package omnia.data.structure.tuple

import java.util.function.Function
import omnia.data.structure.tuple.Tuples.AtMostNonuplet

interface Nonuplet<T : Any> : Nonuple<T, T, T, T, T, T, T, T, T>, AtMostNonuplet<T> {

  override fun <R : Any> map(mapper: Function<in T, out R>): Nonuplet<R>
  override fun dropFirst(): Octuplet<T>
  override fun dropSecond(): Octuplet<T>
  override fun dropThird(): Octuplet<T>
  override fun dropFourth(): Octuplet<T>
  override fun dropFifth(): Octuplet<T>
  override fun dropSixth(): Octuplet<T>
  override fun dropSeventh(): Octuplet<T>
  override fun dropEighth(): Octuplet<T>
  override fun dropNinth(): Octuplet<T>
  override fun concat(`object`: T): Decuplet<T>

  companion object {

    fun <T : Any> of(
        first: T,
        second: T,
        third: T,
        fourth: T,
        fifth: T,
        sixth: T,
        seventh: T,
        eighth: T,
        ninth: T
    ): Nonuplet<T> {
      return ImmutableNonuplet(first, second, third, fourth, fifth, sixth, seventh, eighth, ninth)
    }
  }
}